### PR TITLE
[SPARK-20463] Add support for IS [NOT] DISTINCT FROM.

### PR DESCRIPTION
**Pull request by ptkool**
## What changes were proposed in this pull request?

Add support for the SQL standard distinct predicate to SPARK SQL.

```
<expression> IS [NOT] DISTINCT FROM <expression>
```

## How was this patch tested?

Tested using unit tests, integration tests, manual tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Hm, is this a duplicate of https://github.com/apache/spark/pull/17605 ? I think supporting `eqNullSafe` is more reasonable similarly in other APIs.

File: python/pyspark/sql/column.py [hunk 1]
```diff
@@ -364,6 +364,22 @@ def isin(self, *cols):
     isNull = ignore_unicode_prefix(_unary_op("isNull", _isNull_doc))
     isNotNull = ignore_unicode_prefix(_unary_op("isNotNull", _isNotNull_doc))
 
+    def isNotDistinctFrom(self, col):
```
Comment (gatorsmile) [inline python/pyspark/sql/column.py]: Actually, I like this name. `IS [NOT] DISTINCT FROM` is part of ANSI SQL. We should add alias in Spark SQL too.

**Comment by HyukjinKwon** [comment]
@ptkool Then, I think we should add them all here after fixing description/title in JIRA/PR or close this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -962,6 +962,11 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
         IsNotNull(e)
       case SqlBaseParser.NULL =>
         IsNull(e)
+      case SqlBaseParser.DISTINCT =>
+        ctx.NOT match {
+          case null => Not(EqualTo(e, expression(ctx.right)))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]:
```Scala
case null => Not(EqualTo(e, expression(ctx.right)))
```
->
```Scala
case null => Not(EqualNullSafe(e, expression(ctx.right)))
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 3]
```diff
@@ -167,6 +167,11 @@ class ExpressionParserSuite extends PlanTest {
     assertEqual("a = b is not null", ('a === 'b).isNotNull)
   }
 
+  test("is distinct expressions") {
+    assertEqual("a is distinct from b", !('a === 'b))
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: `!('a === 'b)` -> `!('a <=> 'b)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: BTW, Could you please add a comment in [line 931](https://github.com/ptkool/spark/blob/faf5ebe4f3fad74e6b44de63c8314a853b895e55/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L937) regarding this new syntax support?

Nit: could you please follow how we deal with `IS NOT NULL`?
```Scala
  case SqlBaseParser.DISTINCT if ctx.NOT != null =>
    EqualNullSafe(e, expression(ctx.right))
  case SqlBaseParser.DISTINCT =>
    Not(EqualNullSafe(e, expression(ctx.right)))
```

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by HyukjinKwon** [comment]
I filed [SPARK-20552](https://issues.apache.org/jira/browse/SPARK-20552) for column APIs in both Python and Scala/Java. Do you mind if I work on this if any of you, @gatorsmile and @ptkool, is currently not?

**Comment by redbeardmeric** [comment]
Question:  has this function now been removed from the live build?

**Pull request closed without merging**